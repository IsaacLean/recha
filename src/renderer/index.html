<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <%= csp %>
    <title>Todo List</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>