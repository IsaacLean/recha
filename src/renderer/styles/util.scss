.util-cursor-grab {
  cursor: grab !important;
}

.util-cursor-pointer {
  cursor: pointer !important;
}

.util-display-block {
  display: block !important;
}

.util-display-none {
  display: none !important;
}
