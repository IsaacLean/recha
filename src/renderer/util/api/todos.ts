/* eslint-disable import/prefer-default-export */
import fetch from 'cross-fetch'

import HTTPErr from '../HTTPErr'
import Todo from '../../types/Todo'
import { APIRes } from '../../types'
import { genReqOptions } from '.'

interface TodosQuery {
  col: 'id' | 'date' | 'order_num'
  date: string
  dir: 'ASC' | 'asc' | 'DESC' | 'desc'
}

// @ts-ignore
const TODOS_PATH = `${__API__}/todos` // eslint-disable-line no-undef

export const deleteTodo = (id: number): Promise<APIRes<Todo>> =>
  fetch(`${TODOS_PATH}/${id}`, genReqOptions('DELETE')).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })

export const getTodoLists = (): Promise<APIRes<string[]>> =>
  fetch(`${TODOS_PATH}/lists`).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })

export const getTodos = (query?: Partial<TodosQuery>): Promise<APIRes<Todo[]>> => {
  let newPathTodos = TODOS_PATH
  if (query) {
    const queryString = new URLSearchParams(query).toString()
    if (queryString) newPathTodos += `?${queryString}`
  }

  return fetch(newPathTodos).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })
}

export const patchTodo = (id: number, data: Partial<Todo>): Promise<APIRes<Todo>> =>
  fetch(`${TODOS_PATH}/${id}`, genReqOptions('PATCH', data)).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })

export const patchTodoOrders = (data: number[]): Promise<APIRes<Todo[]>> =>
  fetch(`${TODOS_PATH}/reorder`, genReqOptions('PATCH', data)).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })

export const postTodo = (data: Partial<Todo>): Promise<APIRes<Todo>> =>
  fetch(`${TODOS_PATH}`, genReqOptions('POST', data)).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })
