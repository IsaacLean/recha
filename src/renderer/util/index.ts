/* eslint-disable import/prefer-default-export */
export const todoIsChecked = (completed_at: string): boolean => {
  if (completed_at) return true
  return false
}
