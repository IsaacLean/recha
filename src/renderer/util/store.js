import { applyMiddleware, compose, createStore } from 'redux'

export const setupStore = (reducer, preloadedState, middlewares = [], devMiddlewares) => {
  if(!middlewares) middlewares = []

  if(process.env.NODE_ENV === 'development' && Array.isArray(devMiddlewares)) {
    middlewares = [...middlewares, ...devMiddlewares]
  }

  const composeEnhancers =  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
  if(preloadedState !== null && preloadedState !== undefined) {
    return createStore(reducer, preloadedState, composeEnhancers(applyMiddleware(...middlewares)))
  }
  return createStore(reducer, composeEnhancers(applyMiddleware(...middlewares)))
}
