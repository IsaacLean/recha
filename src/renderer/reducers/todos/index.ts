import { combineReducers } from 'redux'

import api from './api'
import ui from './ui'

const todos = combineReducers({ api, ui })

export default todos
