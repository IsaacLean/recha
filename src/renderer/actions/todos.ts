import { normalize } from 'normalizr'

import Todo from '../types/Todo'
import { APIRes } from '../types'
import { arrayOfTodos } from './schema'
import { getTodos } from '../util/api'
import {
  ActionFetchTodosFailure,
  ActionFetchTodosReq,
  ActionFetchTodosSuccess,
  ActionSelectTodo,
  ActionUpdateFormTodoName,
  FETCH_TODOS_FAILURE,
  FETCH_TODOS_REQ,
  FETCH_TODOS_SUCCESS,
  SELECT_TODO,
  UPDATE_FORM_TODO_NAME,
  NormalizedTodosRes,
  ThunkDispatch,
  ThunkResult
} from '../types/actions'
import { State } from '../types/reducers'

export const fetchTodosFailure = (message: string = 'Something went wrong.'): ActionFetchTodosFailure => ({
  type: FETCH_TODOS_FAILURE,
  message
})

export const fetchTodosReq = (): ActionFetchTodosReq => ({
  type: FETCH_TODOS_REQ
})

export const fetchTodosSuccess = (date: Todo['date'], res: APIRes<Todo[]>): ActionFetchTodosSuccess => {
  return {
    type: FETCH_TODOS_SUCCESS,
    date,
    res: normalize(res.data, arrayOfTodos)
  }
}

export const fetchTodos = (date: Todo['date']): ThunkResult<Promise<ActionFetchTodosFailure | NormalizedTodosRes>> => (
  dispatch: ThunkDispatch,
  getState: () => State
) => {
  // if (getIsFetching(getState(), filter)) {
  //   return Promise.resolve()
  // }

  dispatch(fetchTodosReq())

  return getTodos({ date, col: 'order_num', dir: 'ASC' }).then(
    res => Promise.resolve(dispatch(fetchTodosSuccess(date, res)).res),
    err => Promise.resolve(dispatch(fetchTodosFailure(err.message)))
  )
}

export const selectTodo = (id: Todo['id'], name: Todo['name']): ActionSelectTodo => ({
  type: SELECT_TODO,
  id,
  name
})

export const updateFormTodoName = (id: Todo['id'], name: Todo['name']): ActionUpdateFormTodoName => ({
  type: UPDATE_FORM_TODO_NAME,
  id,
  name
})
