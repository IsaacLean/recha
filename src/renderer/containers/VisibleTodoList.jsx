import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import FetchError from '../components/FetchError'
import TodoList from '../components/TodoList'
import { fetchTodos, toggleTodo } from '../actions'
import { getErrorMessage, getIsFetching, getVisibleTodos } from '../reducers'

class VisibleTodoList extends Component {
  componentDidMount() {
    this.fetchData()
  }

  componentDidUpdate(prevProps) {
    if(this.props.filter !== prevProps.filter) {
      this.fetchData()
    }
  }

  fetchData() {
    const { filter, fetchTodos } = this.props
    fetchTodos(filter)
  }

  render() {
    const { errorMessage, isFetching, toggleTodo, todos } = this.props

    if(isFetching && !todos.length) {
      return <p>Loading...</p>
    }

    if(errorMessage && !todos.length) {
      return (
        <FetchError
          message={errorMessage}
          onRetry={() => this.fetchData()}
        />
      )
    }

    return <TodoList
      todos={todos}
      onTodoClick={toggleTodo}
    />
  }
}

VisibleTodoList.propTypes = {
  errorMessage: PropTypes.string,
  fetchTodos: PropTypes.func.isRequired,
  filter: PropTypes.string.isRequired,
  isFetching: PropTypes.bool.isRequired,
  todos: PropTypes.arrayOf(PropTypes.object),
  toggleTodo: PropTypes.func.isRequired
}

const mapStateToProps = (state, { match }) => {
  const filter = match.params.filter || 'all'

  return {
    errorMessage: getErrorMessage(state, filter),
    filter,
    isFetching: getIsFetching(state, filter),
    todos: getVisibleTodos(state, filter)
  }
}

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchTodos,
  toggleTodo
}, dispatch)

const VisibleTodoListContainer = withRouter(connect(
  mapStateToProps,
  mapDispatchToProps
)(VisibleTodoList))

export default VisibleTodoListContainer
