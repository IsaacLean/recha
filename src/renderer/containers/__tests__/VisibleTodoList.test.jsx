import React from 'react'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'
import { HashRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux'

import todos from '../../reducers'
import VisibleTodoList from '../VisibleTodoList'
import { setupStore } from '../../util/store'

describe('VisibleTodoList', () => {
  it('renders properly', () => {
    const container = renderer.create(
      <Provider store={setupStore(todos, null, [thunk])}>
        <Router>
          <VisibleTodoList />
        </Router>
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
