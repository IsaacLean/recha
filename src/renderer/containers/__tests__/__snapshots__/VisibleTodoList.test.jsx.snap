// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisibleTodoList renders properly 1`] = `
<p>
  Loading...
</p>
`;
